ethash/src/keccak.rs

```diff
@@ -21,36 +21,27 @@ pub type H256 = [u8; 32];
 pub mod keccak_512 {
 	use super::hash;
 
-	pub use self::hash::keccak_512_unchecked as unchecked;
+	pub use self::hash::{keccak_512 as write, keccak512 as inplace};
 
-	pub fn write(input: &[u8], output: &mut [u8]) {
-		hash::keccak_512(input, output);
-	}
+	pub unsafe fn unchecked(out: *mut u8, outlen: usize, input: *const u8, inputlen: usize) {
+		// TODO(ordian): is this UB?
+		let src = core::slice::from_raw_parts(input, inputlen);
+		let dst = core::slice::from_raw_parts_mut(out, outlen);
 
-	pub fn inplace(input: &mut [u8]) {
-		// This is safe since `keccak_*` uses an internal buffer and copies the result to the output. This
-		// means that we can reuse the input buffer for both input and output.
-		unsafe {
-			hash::keccak_512_unchecked(input.as_mut_ptr(), input.len(), input.as_ptr(), input.len());
-		}
+		self::hash::keccak_512(src, dst);
 	}
 }
 
 pub mod keccak_256 {
 	use super::hash;
 
-	pub use self::hash::keccak_256_unchecked as unchecked;
+	pub use self::hash::{keccak_256 as write, keccak256 as inplace};
 
-	#[allow(dead_code)]
-	pub fn write(input: &[u8], output: &mut [u8]) {
-		hash::keccak_256(input, output);
-	}
+	pub unsafe fn unchecked(out: *mut u8, outlen: usize, input: *const u8, inputlen: usize) {
+		// TODO(ordian): is this UB?
```

Comment:
//cc @ordian 

Can you explain why this might be UB? 

What I can think of is:

- Deref `null`
- Either the one of the pointers are unaligned
- `out` and `input` are overlapping

Reply:
This method needs to have `safety documentation` in order to instruct how to use this API properly.

Reply:
https://doc.rust-lang.org/std/slice/fn.from_raw_parts_mut.html
> The memory referenced by the returned slice must not be accessed through any other pointer (not derived from the return value) for the duration of lifetime 'a. Both read and write accesses are forbidden.

Reply:
and out and input are definitely overlapping (that's the use case)

Reply:
Ok, I see thanks, didn't read the code properly.